### PR TITLE
User home directory Wine environments and additional distribution support.

### PR DESCRIPTION
**Pull request by gordonrs**
Changes I proposed to @ChrisTruncer last week via email concerning the installer. This is the first PR from that proposal.

The premise behind this was simply to begin to allow non-root users be able to run Veil Evasion and add additional distribution support. This required setting up wine environments inside the user's home directory and allowing the wine environment to be nuked from the installer if a previous environment is found. This required segregating off the Veil environment from one that a user might previously have on their system and we wouldn't want to interfere with it. As such, the location for this is now ~/.config/wine/veil/.

Additional distribution support has been opened up with this PR. Veil-Evasion installation now works on Arch Linux, Debian 8, Debian sid/testing and Ubuntu 15.10. Additional distribution support will be added by me on request if people really need it. The current Fedora support in there will get shored up in a future PR if people request it.

Another future PR that I have planned will get the installer filesystem spew under control and make sure everything stays within ~/.config/wine/veil.

The changes have been tested on the following distributions: Kali Linux 2, Debian 8, Ubuntu 15.10 and Arch Linux. Currently it installs and runs on all of those platforms. Some simple binaries have been put through Veil by myself but it would awesome if some other folks could help me test that these changes won't break anything for anyone's day-to-day operation.

I highly welcome comments and suggestions on how to improve this further. You'll catch me idling in #Veil on IRC.

### CODE REVIEW

**Comment by ChrisTruncer** [comment]
Hey, this is awesome work.  Thanks for sending this over to me.  I'm working on downloading a clean copy of Kali to test everything on at the moment, but I'm in a hotel and the wifi isn't the best.  Once it's all downloaded I'll be sure to get testing this right away.

**Comment by necrose99** [comment]
I had mentioned much the same with some gentoo provision wine scripts
Ie
$user/.veil-wine..

Gentoo ebuilds is more a pain without a setup.py or makefile.
Bash ebuilds at presently expect setup.py,
Normaly builds in var sandbox, makes disdir... then tests , then moves to
live fs.  , rigging the dirs and forcing distdir is a pain.  Point being if
each module deploys to /usr/bin/veil.... or the like
Veil-?-9999 ie github live.  Else stable version numbers.
Though I can unpack debs via ebuilds.. tends to be frowned on at times.

I have some draft scripts for wine tricks and wget packages. In
necrose99/necromancy...

Also had a draft to shred the directory's files with file shreader.  After
snapping logs or etc or not .. shread it clean.

In middle of cissp prep so I've my hands full.

File: setup/setup.sh [hunk 1]
```diff
@@ -2,8 +2,25 @@
 
 
 # Global Variables
+arch="$(uname -m)"
 runuser="$(whoami)"
+trueuser="$(who am i | awk '{print $1}')" # if this is blank, we're actually root (kali)
```
Comment (void-in) [inline setup/setup.sh]: Instead of ```$(who am i | awk '{print $1)```, we can simply call ```$(whoami)``` and it will return only the username. No need to use awk.

**Comment by gordonrs** [comment]
@ChrisTruncer No problem! Hopefully there's no issues at runtime resulting from the setup. I'm going to wait until this is signed off on before addressing the filesystem spray. Speaking of, is there anything special about the version of pyinstaller provided by Veil that would prevent it from being installed via pip?

@necrose99 I haven't touched a Gentoo box in 12 years so I wouldn't hazard a guess on how to make the installation on Gentoo easier. My best bet would be to say, just clone the git repo and benefit from the other changes that is being made. I don't know if it's possible to set it up like you can with Arch packages (that I plan on providing through the AUR) by just having the installation package clone the git repo and running setup/setup.sh and letting it do it's thing. After that, just setup symbolic links to wherever Veil binaries are installed on the system, but that's for later planning. Main goal of this PR is to just segregate the wine environments.

**Comment by ChrisTruncer** [comment]
@gordonrs - at home, download clean iso, and will be playing with it.  Just wanted to make sure you knew I haven't forgotten about this.

**Comment by gordonrs** [comment]
No problem! Let me know how it goes when you have time. If there's any issues, I'll get them sorted out.

**Comment by gordonrs** [comment]
Just a quick update. It looks like I'll also have to change the hardcoded paths in modules/common/supportfiles.py for the wine directories that I had overlooked. Chris, could you hold off on this PR until I can make those changes?

Thanks.

**Comment by ChrisTruncer** [comment]
Yeah,  No problem.

Working on getting clean environment setup as we speak actually, so that works.

Again, thanks for your help with this!


-----------------------------------------------------------------------------------------
Christopher Truncer
@christruncer

On March 7, 2016 at 8:03:03 AM, Gordon Robert Speirs (notifications@github.com) wrote:

Just a quick update. It looks like I'll also have to change the hardcoded paths in modules/common/supportfiles.py for the wine directories that I had overlooked. Chris, could you hold off on this PR until I can make those changes?

Thanks.

**Comment by gordonrs** [comment]
It's no problem mate, happy to help out!

**Comment by ChrisTruncer** [comment]
Also, just let me know once you've pushed the change, so I can check it out locally here

**Comment by gordonrs** [comment]
_Sorry for the last two commits there being a little bit of a mess, had to comment stuff out and so forth when testing and forgot to clean that up. Hence the second commit_

This is now good to go. I've been doing extensive testing on it and everything now seems to work. The previous issue that I mentioned on IRC with ruby payloads not generating. That was an issue on my end. If a user is using the ruby version manager (as I was for some of my JRuby work) then it causes some environment weirdness. This is easily fixed by telling rvm to use the system "rvm system".

It has been tested on the following distributions:

- Kali Rolling
- Debian 8
- Ubuntu 15.10
- Arch Linux

All were able to generate payloads, work fine and have working sessions. I encourage further testing, though I've been testing this for the last week and the odd corner case issue that have cropped up have been eliminated. In particular, it's best if we explicitly state that we're using python 2 version of things that prevents pip package version weirdness in Ubuntu and Arch.

**Comment by ChrisTruncer** [comment]
Hey, I just started looking at this.  Looks good so far, I found a bug, but I it's my original bug (not yours).  

Line 86 within the letter_substitution python module.  It should by payload_code += randctypes .... vs payload_code += ranctypes.  This is something I will need to change, but if you just want to add it in, I can get it merged with your push (vs. you having to re-sync with my fix for it).

I'm continuing to go through a play with this, so thanks!

**Comment by gordonrs** [comment]
No problem, fixed that up for you. :)

**Comment by ChrisTruncer** [comment]
@gordonrs I just got done testing this, and really nice work.  I absolutely appreciate the help in cleaning this code up, and it was done really well.

Thanks, and I'll get this merged in!

**Comment by necrose99** [comment]
@gordonrs @ChrisTruncer
I use [Sabayon](url) /
https://atlas.hashicorp.com/Sabayon/boxes/spinbase-amd64 
Pentoo Spike , etc  overlay Mainly its Gentoo for the most part , it just has a bin-package manager , goes on easier and less hell.. don't have to edit configs for weeks.. 
and Gentoo is just under the hood.  for easy dev-harking  when you need to build shit that is. 
gcc-config .... or eselect "dev-toy" wanted ie automake and pick a version... , makes also testing or legacy apps support also easier. 

 anyhow ** Segregation of Wine Environment  was on my list also.**   **distro agnostic setups also**
and just packaging a script for Gentoo to spec out wine specifically for use with veil , well its upstream now so much easier....  can just call the shell for it. or just install it via ebuild. or merely adapt it.  Thanks...

setup.sh , Veil-*Atoms , setup.py setup.cfg ,  (of which later do-RPM , DO-DEB can be added if liked (makes RPM or deb's for Install ) 
but in essence /usr/bin/veil* , /opt/usr/veil* installs (will keep official path) , setup symlinks.  ebuild (bash & Eclass will just run veil-evasion setup.py , thus just putting individual items in proper place inside build root then copied into live-file-system  symlinks etc.  to /usr/bin , fixperms ie chmod+x 

likewise for veil-catapult setup.py (depend on veil-evasion etc) 

setup.sh , then just call git and python setup.py from each , its more universal and can make pip installs etc. eggs or what have you.  

**For any-other distro this also will make package building easier**. ie Pkg-build say arch. 
or similar os. OSX , BSD ... 
call /veil-*  setup.py , also adds in each item modularaly  ,  can add a not to ebuilds main config is ...
need to edit read install directions @ veil-framwork etc. or include /usr/share/docs for usual suspects. 
 
bonus is https://ghantoos.org/2008/10/19/creating-a-deb-package-from-a-python-setuppy/ 
https://pypi.python.org/pypi/stdeb/0.8.5
<u>which only helps QA  on main project</u> , but also keeps  it simple on other distros. and adds less downstream support issues to slog though. 

https://github.com/necrose99/necromancy-overlay/blob/master/net-analyzer/veil/veil-meta-0.7.0-r1.ebuild  (dose basically the same as Veil-framework setup.sh however it calls more ebuilds to do each setup as a depend &  its modular)  

https://github.com/necrose99/necromancy-overlay/blob/master/app-emulation/veil-wine/veil-wine.ebuild  needing to simplify and or correct wine paths since support for user wine prefix is now upstream.  64bit was for some tinkering only and not advised for normal use...  but basically wget , and run into wine-prefix dir to setup crap. however if a specified x64 test case path is up-streamed to veil-framework I will follow it , for now I'll include into files but wont have it run-able for install.  

https://github.com/necrose99/necromancy-overlay/blob/master/net-analyzer/veil-evasion/veil-evasion-2.21.1.1.ebuild  if their were veil-evasion setup.py/setup.cfg to call it would do so...
9999 version denotes github latest commit  forcing DESTDIR is a bitch.  

when I get a chance i'm going to Add in setup.py and setup.cfg on a branch then push back 
basic install symlinks to /usr/bin , system install ie /usr/bin/veil-framwork or /usr/opt/bin/veil-framework/?dirs    if can be made stable , then it will be more distro agnostic. and can be merged into codebase at such time. 

anyone elase makes EGGS PIP install , make RPM or a DEB more power to em. , just going for relay Generic setup.py's  to do this , solves @Gentoo @Pentoo @Sabayon @Funtoo  & Gentoo related Install Issues , and for that fact likely others.. 
others can Amend latter setup.cfg etc after the fact.

**Comment by gordonrs** [comment]
I certainly can't speak for @ChrisTruncer or the rest of the devs on Veil on this, but my input would be the following:

Packaging should be done by the distributions and not by upstream. It would add way too much of a support overhead to the handful of folks where there are other things to be done with Veil. That's just my opinion on it. 

The job I'm taking on is making it easier to enable distro packagers to actually do this. That includes "normalizing" the paths that we use, where stuff is installed and so on. The wine environments in this PR were the first step that enables the rest of the work to be taken on. 

I still have more work to do on this and work on it is slow because I'm swamped with work and have an election campaign here in Scotland that is doing to drain pretty much all of my effort over the next 5 weeks.

Basically, it would be best if our setup process was as generic as possible, but I don't think adding in distro specific installers would actually be the way to go on this. I'm certainly not going to be including the Arch PKGBUILD upstream because of this issue. It's not required. This is the internal fight I've had with myself regarding OS X support also. It would be too much to change for whenever we want to change additional things in the future.

I don't make the calls on it though, it's entirely up to Chris and the guys where they'd want to go with it.

**Comment by ChrisTruncer** [comment]
I agree with @gordonrs.  We're not in the business of packaging or forcing maintainers to use our package.  We'll let them do it in a way that best works with their distro.
